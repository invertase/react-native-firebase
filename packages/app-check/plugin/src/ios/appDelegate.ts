import { ConfigPlugin, IOSConfig, WarningAggregator, withDangerousMod } from '@expo/config-plugins';
import { AppDelegateProjectFile } from '@expo/config-plugins/build/ios/Paths';
import { mergeContents } from '@expo/config-plugins/build/utils/generateCode';
import fs from 'fs';

const methodInvocationBlock = `[RNFBAppCheckModule sharedInstance];`;
// https://regex101.com/r/mPgaq6/1
const methodInvocationLineMatcher =
  /(?:self\.moduleName\s*=\s*@\"([^"]*)\";)|(?:(self\.|_)(\w+)\s?=\s?\[\[UMModuleRegistryAdapter alloc\])|(?:RCTBridge\s?\*\s?(\w+)\s?=\s?\[(\[RCTBridge alloc\]|self\.reactDelegate))/g;

// https://regex101.com/r/nHrTa9/1/
// if the above regex fails, we can use this one as a fallback:
const fallbackInvocationLineMatcher =
  /-\s*\(BOOL\)\s*application:\s*\(UIApplication\s*\*\s*\)\s*\w+\s+didFinishLaunchingWithOptions:/g;

export function modifyObjcAppDelegate(contents: string): string {
  // Add import
  if (!contents.includes('#import <RNFBAppCheckModule.h>')) {
    contents = contents.replace(
      /#import "AppDelegate.h"/g,
      `#import "AppDelegate.h"
#import <RNFBAppCheckModule.h>`,
    );
  }

  // To avoid potential issues with existing changes from older plugin versions
  if (contents.includes(methodInvocationBlock)) {
    return contents;
  }

  if (
    !methodInvocationLineMatcher.test(contents) &&
    !fallbackInvocationLineMatcher.test(contents)
  ) {
    WarningAggregator.addWarningIOS(
      '@react-native-firebase/app-check',
      'Unable to determine correct Firebase insertion point in AppDelegate.m. Skipping Firebase addition.',
    );
    return contents;
  }

  // Add invocation
  try {
    return mergeContents({
      tag: '@react-native-firebase/app-check-didFinishLaunchingWithOptions',
      src: contents,
      newSrc: methodInvocationBlock,
      anchor: methodInvocationLineMatcher,
      offset: 0, // new line will be inserted right above matched anchor
      comment: '//',
    }).contents;
  } catch (_: any) {
    // tests if the opening `{` is in the new line
    const multilineMatcher = new RegExp(fallbackInvocationLineMatcher.source + '.+\\n*{');
    const isHeaderMultiline = multilineMatcher.test(contents);

    // we fallback to another regex if the first one fails
    return mergeContents({
      tag: '@react-native-firebase/app-didFinishLaunchingWithOptions-fallback',
      src: contents,
      newSrc: methodInvocationBlock,
      anchor: fallbackInvocationLineMatcher,
      // new line will be inserted right below matched anchor
      // or two lines, if the `{` is in the new line
      offset: isHeaderMultiline ? 2 : 1,
      comment: '//',
    }).contents;
  }
}

export function modifySwiftAppDelegate(contents: string): string {
  // Add imports for Swift
  if (!contents.includes('import RNFBAppCheck')) {
    // Try to add after FirebaseCore if it exists
    if (contents.includes('import FirebaseCore')) {
      contents = contents.replace(
        /import FirebaseCore/g,
        `import FirebaseCore
import RNFBAppCheck`,
      );
    } else {
      // Otherwise add after Expo
      contents = contents.replace(
        /import Expo/g,
        `import Expo
import RNFBAppCheck`,
      );
    }
  }

  // Check if App Check code is already added to avoid duplication
  if (contents.includes('RNFBAppCheckModule.sharedInstance()')) {
    return contents;
  }

  // Find the Firebase initialization end line to insert after
  const firebaseLine = '// @generated end @react-native-firebase/app-didFinishLaunchingWithOptions';

  if (contents.includes(firebaseLine)) {
    // Insert right after Firebase initialization
    return contents.replace(
      firebaseLine,
      `${firebaseLine}
        FirebaseApp.configure()
        RNFBAppCheckModule.sharedInstance()
      `,
    );
  }

  // If Firebase initialization block not found, add both Firebase and App Check initialization
  // This is to make sure Firebase is initialized before App Check
  const methodInvocationBlock = `FirebaseApp.configure()
    RNFBAppCheckModule.sharedInstance()`;

  const methodInvocationLineMatcher = /(?:factory\.startReactNative\()/;

  if (!methodInvocationLineMatcher.test(contents)) {
    WarningAggregator.addWarningIOS(
      '@react-native-firebase/app-check',
      'Unable to determine correct insertion point in AppDelegate.swift. Skipping App Check addition.',
    );
    return contents;
  }

  try {
    return mergeContents({
      tag: '@react-native-firebase/app-check',
      src: contents,
      newSrc: methodInvocationBlock,
      anchor: methodInvocationLineMatcher,
      offset: 0,
      comment: '//',
    }).contents;
  } catch (_e) {
    WarningAggregator.addWarningIOS(
      '@react-native-firebase/app-check',
      'Failed to insert App Check initialization code.',
    );
    return contents;
  }
}

export async function modifyAppDelegateAsync(appDelegateFileInfo: AppDelegateProjectFile) {
  const { language, path, contents } = appDelegateFileInfo;

  let newContents;
  if (['objc', 'objcpp'].includes(language)) {
    newContents = modifyObjcAppDelegate(contents);
  } else if (language === 'swift') {
    newContents = modifySwiftAppDelegate(contents);
  } else {
    throw new Error(`Cannot add Firebase code to AppDelegate of language "${language}"`);
  }

  await fs.promises.writeFile(path, newContents);
}

export const withFirebaseAppDelegate: ConfigPlugin = config => {
  return withDangerousMod(config, [
    'ios',
    async config => {
      const fileInfo = IOSConfig.Paths.getAppDelegate(config.modRequest.projectRoot);
      await modifyAppDelegateAsync(fileInfo);
      return config;
    },
  ]);
};
