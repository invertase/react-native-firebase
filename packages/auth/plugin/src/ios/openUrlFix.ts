import {
  ConfigPlugin,
  withAppDelegate,
  withInfoPlist,
  ExportedConfigWithProps,
  WarningAggregator,
} from '@expo/config-plugins';
import type { ExpoConfig } from '@expo/config/build/Config.types';
import type { AppDelegateProjectFile } from '@expo/config-plugins/build/ios/Paths';
import type { InfoPlist } from '@expo/config-plugins/build/ios/IosConfig.types';
import { mergeContents } from '@expo/config-plugins/build/utils/generateCode';
import { PluginConfigType } from '../pluginConfig';

export const withIosCaptchaOpenUrlFix: ConfigPlugin<PluginConfigType> = (
  config: ExpoConfig,
  props?: PluginConfigType,
) => {
  // check configuration
  if (!shouldApplyIosOpenUrlFix({ config, props })) {
    return config;
  }

  // check that swizzling is enabled; otherwise patch must be customized and applied manually
  withInfoPlist(config, config => {
    ensureFirebaseSwizzlingIsEnabled(config);
    return config;
  });

  // apply patch
  return withAppDelegate(config, config => {
    return withOpenUrlFixForAppDelegate({ config, props });
  });
};

// Interpret the plugin config to determine whether this fix should be applied
export function shouldApplyIosOpenUrlFix({
  config,
  props,
}: {
  config: ExpoConfig;
  props?: PluginConfigType;
}): boolean {
  const flag = props?.ios?.captchaOpenUrlFix;
  if (flag === undefined || flag === 'default') {
    // by default, apply the fix whenever 'expo-router' is detected in the same project
    return isPluginEnabled(config, 'expo-router');
  } else if (flag === true || flag === false) {
    const isEnabled: boolean = flag;
    return isEnabled;
  } else {
    throw new Error(`Unexpected value for 'captchaOpenUrlFix' config option`);
  }
}

export function withOpenUrlFixForAppDelegate({
  config,
  props,
}: {
  config: ExportedConfigWithProps<AppDelegateProjectFile>;
  props?: PluginConfigType;
}) {
  const { language, contents } = config.modResults;

  if (['objc', 'objcpp'].includes(language)) {
    const newContents = modifyObjcAppDelegate(contents);
    if (newContents === null) {
      if (props?.ios?.captchaOpenUrlFix === true) {
        throw new Error("Failed to apply iOS openURL fix because no 'openURL' method was found");
      } else {
        WarningAggregator.addWarningIOS(
          '@react-native-firebase/auth',
          "Skipping iOS openURL fix because no 'openURL' method was found",
        );
        return config;
      }
    } else {
      return {
        ...config,
        modResults: {
          ...config.modResults,
          contents: newContents,
        },
      };
    }
  } else {
    // TODO: Support Swift
    throw new Error(`Don't know how to apply openUrlFix to AppDelegate of language "${language}"`);
  }
}

const skipOpenUrlForFirebaseAuthBlock: string = `\
  if ([url.host caseInsensitiveCompare:@"firebaseauth"] == NSOrderedSame) {
    // invocations for Firebase Auth are handled elsewhere and should not be forwarded to Expo Router
    return NO;
  }\
`;

// NOTE: `mergeContents()` requires that this pattern not match newlines
export const appDelegateOpenUrlInsertionPointAfter: RegExp =
  /-\s*\(\s*BOOL\s*\)\s*application\s*:\s*\(\s*UIApplication\s*\*\s*\)\s*application\s+openURL\s*:\s*\(\s*NSURL\s*\*\s*\)\s*url\s+options\s*:\s*\(\s*NSDictionary\s*<\s*UIApplicationOpenURLOptionsKey\s*,\s*id\s*>\s*\*\s*\)\s*options\s*/; // ðŸ™ˆ

export const multiline_appDelegateOpenUrlInsertionPointAfter = new RegExp(
  appDelegateOpenUrlInsertionPointAfter.source + '\\s*{\\s*\\n',
);

// Returns AppDelegate with modification applied, or null if no change needed.
export function modifyObjcAppDelegate(contents: string): string | null {
  const pattern = appDelegateOpenUrlInsertionPointAfter;
  const multilinePattern = multiline_appDelegateOpenUrlInsertionPointAfter;
  const fullMatch = contents.match(multilinePattern);
  if (!fullMatch) {
    if (contents.match(pattern)) {
      throw new Error("Failed to find insertion point; expected newline after '{'");
    } else if (contents.match(/openURL\s*:/)) {
      throw new Error(
        [
          "Failed to apply 'captchaOpenUrlFix' but detected 'openURL' method.",
          "Please manually apply the fix to your AppDelegate's openURL method,",
          "then update your app.config.json by configuring the '@react-native-firebase/auth' plugin",
          'to set `captchaOpenUrlFix: false`.',
        ].join(' '),
      );
    } else {
      return null;
    }
  }
  const fullMatchNumLines = fullMatch[0].split('\n').length;
  const offset = fullMatchNumLines - 1;
  if (offset < 0) {
    throw new Error(`Failed to find insertion point; fullMatchNumLines=${fullMatchNumLines}`);
  }
  return mergeContents({
    tag: '@react-native-firebase/auth-openURL',
    src: contents,
    newSrc: skipOpenUrlForFirebaseAuthBlock,
    anchor: appDelegateOpenUrlInsertionPointAfter,
    offset,
    comment: '//',
  }).contents;
}

export type ExpoConfigPluginEntry = string | [] | [string] | [string, any];

// Search the ExpoConfig plugins array to see if `pluginName` is present
function isPluginEnabled(config: ExpoConfig, pluginName: string): boolean {
  if (config.plugins === undefined) {
    return false;
  }
  return config.plugins.some((plugin: ExpoConfigPluginEntry) => {
    if (plugin === pluginName) {
      return true;
    } else if (Array.isArray(plugin) && plugin.length >= 1 && plugin[0] === pluginName) {
      return true;
    } else {
      return false;
    }
  });
}

export function isFirebaseSwizzlingDisabled(config: ExportedConfigWithProps<InfoPlist>): boolean {
  return config.ios?.infoPlist?.['FirebaseAppDelegateProxyEnabled'] === false;
}

export function ensureFirebaseSwizzlingIsEnabled(
  config: ExportedConfigWithProps<InfoPlist>,
): boolean {
  if (isFirebaseSwizzlingDisabled(config)) {
    throw new Error(
      [
        'Your app has disabled swizzling by setting FirebaseAppDelegateProxyEnabled=false in its Info.plist.',
        "Please update your app.config.json to configure the '@react-native-firebase/auth' plugin to set `captchaOpenUrlFix: false`",
        'and see https://firebase.google.com/docs/auth/ios/phone-auth#appendix:-using-phone-sign-in-without-swizzling for instructions.',
      ].join(' '),
    );
  } else {
    return true;
  }
}
