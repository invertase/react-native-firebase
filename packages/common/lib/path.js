/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

/**
 * Returns the next parent of the path e.g. /foo/bar/car -> /foo/bar
 */
export function pathParent(path) {
  if (path.length === 0) {
    return null;
  }

  const index = path.lastIndexOf('/');
  if (index <= 0) {
    return null;
  }

  return path.slice(0, index);
}

/**
 * Joins a parent and a child path
 */
export function pathChild(path, childPath) {
  const canonicalChildPath = pathPieces(childPath).join('/');

  if (path.length === 0) {
    return canonicalChildPath;
  }

  return `${path}/${canonicalChildPath}`;
}

/**
 * Returns the last component of a path, e.g /foo/bar.jpeg -> bar.jpeg
 */
export function pathLastComponent(path) {
  const index = path.lastIndexOf('/', path.length - 2);
  if (index === -1) {
    return path;
  }

  return path.slice(index + 1);
}

/**
 * Returns all none empty pieces of the path
 * @param path
 * @returns {*}
 */
export function pathPieces(path) {
  return path.split('/').filter($ => $.length > 0);
}

/**
 * Returns whether a given path is empty
 * @param path
 * @returns {boolean}
 */
export function pathIsEmpty(path) {
  return !pathPieces(path).length;
}

/**
 * Converts a given path to a URL encoded string
 * @param path
 * @returns {string|string}
 */
export function pathToUrlEncodedString(path) {
  const pieces = pathPieces(path);
  let pathString = '';
  for (let i = 0; i < pieces.length; i++) {
    pathString += `/${encodeURIComponent(String(pieces[i]))}`;
  }
  return pathString || '/';
}

// eslint-disable-next-line no-control-regex
export const INVALID_PATH_REGEX = /[[\].#$\u0000-\u001F\u007F]/;

/**
 * Ensures a given path is a valid Firebase path
 * @param path
 * @returns {boolean}
 */
export function isValidPath(path) {
  return typeof path === 'string' && path.length !== 0 && !INVALID_PATH_REGEX.test(path);
}

// eslint-disable-next-line no-control-regex
export const INVALID_KEY_REGEX = /[\[\].#$\/\u0000-\u001F\u007F]/;

/**
 * Ensures a given key is a valid Firebase key
 * @param key
 * @returns {boolean}
 */
export function isValidKey(key) {
  return typeof key === 'string' && key.length !== 0 && !INVALID_KEY_REGEX.test(path);
}

/**
 * Converts a file path to a standardized string path
 * @param path
 * @returns {*}
 */
export function toFilePath(path) {
  let _filePath = path.replace('file://', '');
  if (_filePath.includes('%')) {
    _filePath = decodeURIComponent(_filePath);
  }
  return _filePath;
}
