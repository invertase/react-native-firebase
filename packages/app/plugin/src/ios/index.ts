import { withFirebaseAppDelegate } from './appDelegate';
import { withIosGoogleServicesFile } from './googleServicesPlist';

export { withIosGoogleServicesFile, withFirebaseAppDelegate };
