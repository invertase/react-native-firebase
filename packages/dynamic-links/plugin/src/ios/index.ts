import { withFirebaseAppDelegate } from './appDelegate';

export { withFirebaseAppDelegate };
