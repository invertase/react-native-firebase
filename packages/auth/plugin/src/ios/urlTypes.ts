import {
  ConfigPlugin,
  IOSConfig,
  withInfoPlist,
  ExportedConfigWithProps,
} from '@expo/config-plugins';
import fs from 'fs';
import path from 'path';
import plist from 'plist';

// does this for you: https://firebase.google.com/docs/auth/ios/phone-auth#enable-phone-number-sign-in-for-your-firebase-project
export const withIosCaptchaUrlTypes: ConfigPlugin = config => {
  return withInfoPlist(config, config => {
    return setUrlTypesForCaptcha({ config });
  });
};

function getReversedClientId(googleServiceFilePath: string): string {
  try {
    const googleServicePlist = fs.readFileSync(googleServiceFilePath, 'utf8');

    const googleServiceJson = plist.parse(googleServicePlist) as { REVERSED_CLIENT_ID: string };
    const REVERSED_CLIENT_ID = googleServiceJson.REVERSED_CLIENT_ID;

    if (!REVERSED_CLIENT_ID) {
      throw new TypeError('REVERSED_CLIENT_ID missing');
    }

    return REVERSED_CLIENT_ID;
  } catch {
    throw new Error(
      '[@react-native-firebase/auth] Failed to parse your GoogleService-Info.plist. Are you sure it is a valid Info.Plist file with a REVERSE_CLIENT_ID field?',
    );
  }
}

// add phone auth support by configuring recaptcha
// https://github.com/invertase/react-native-firebase/pull/6167
function addUriScheme(
  config: ExportedConfigWithProps<IOSConfig.InfoPlist>,
  reversedClientId: string,
): ExportedConfigWithProps<IOSConfig.InfoPlist> {
  if (!config.modResults) {
    config.modResults = {};
  }

  if (!config.modResults.CFBundleURLTypes) {
    config.modResults.CFBundleURLTypes = [];
  }

  const hasReverseClientId = config.modResults.CFBundleURLTypes?.some(urlType =>
    urlType.CFBundleURLSchemes.includes(reversedClientId),
  );

  if (!hasReverseClientId) {
    config.modResults.CFBundleURLTypes.push({
      CFBundleURLSchemes: [reversedClientId],
    });
  }

  return config;
}

export function setUrlTypesForCaptcha({
  config,
}: {
  config: ExportedConfigWithProps<IOSConfig.InfoPlist>;
}) {
  const googleServicesFileRelativePath = config.ios?.googleServicesFile;
  if (!googleServicesFileRelativePath) {
    throw new Error(
      `[@react-native-firebase/auth] Your app.json file is missing ios.googleServicesFile. Please add this field.`,
    );
  }
  const googleServiceFilePath = path.resolve(
    config.modRequest.projectRoot,
    googleServicesFileRelativePath,
  );

  if (!fs.existsSync(googleServiceFilePath)) {
    throw new Error(
      `[@react-native-firebase/auth] GoogleService-Info.plist doesn't exist in ${googleServiceFilePath}. Place it there or configure the path in app.json`,
    );
  }

  const reversedClientId = getReversedClientId(googleServiceFilePath);
  addUriScheme(config, reversedClientId);

  return config;
}
