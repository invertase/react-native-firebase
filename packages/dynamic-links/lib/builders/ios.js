/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { isObject, isString } from '@react-native-firebase/app/lib/common';

export default function buildIos(iosParams) {
  if (!isObject(iosParams)) {
    throw new Error("'dynamicLinksParams.ios' must be an object.");
  }

  const params = {};

  if (iosParams.appStoreId) {
    if (!isString(iosParams.appStoreId)) {
      throw new Error("'dynamicLinksParams.ios.appStoreId' must be a string.");
    }

    params.appStoreId = iosParams.appStoreId;
  }

  if (iosParams.bundleId) {
    if (!isString(iosParams.bundleId)) {
      throw new Error("'dynamicLinksParams.ios.bundleId' must be a string.");
    }

    params.bundleId = iosParams.bundleId;
  }

  if (iosParams.customScheme) {
    if (!isString(iosParams.customScheme)) {
      throw new Error("'dynamicLinksParams.ios.customScheme' must be a string.");
    }

    params.customScheme = iosParams.customScheme;
  }

  if (iosParams.fallbackUrl) {
    if (!isString(iosParams.fallbackUrl)) {
      throw new Error("'dynamicLinksParams.ios.fallbackUrl' must be a string.");
    }

    params.fallbackUrl = iosParams.fallbackUrl;
  }

  if (iosParams.iPadBundleId) {
    if (!isString(iosParams.iPadBundleId)) {
      throw new Error("'dynamicLinksParams.ios.iPadBundleId' must be a string.");
    }

    params.iPadBundleId = iosParams.iPadBundleId;
  }

  if (iosParams.iPadFallbackUrl) {
    if (!isString(iosParams.iPadFallbackUrl)) {
      throw new Error("'dynamicLinksParams.ios.iPadFallbackUrl' must be a string.");
    }

    params.iPadFallbackUrl = iosParams.iPadFallbackUrl;
  }

  if (iosParams.minimumVersion) {
    if (!isString(iosParams.minimumVersion)) {
      throw new Error("'dynamicLinksParams.ios.minimumVersion' must be a string.");
    }

    params.minimumVersion = iosParams.minimumVersion;
  }

  if (!params.bundleId) {
    throw new Error("'dynamicLinksParams.ios' missing required 'bundleId' property.");
  }

  return params;
}
