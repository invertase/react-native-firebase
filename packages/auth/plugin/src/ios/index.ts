import { withIosCaptchaUrlTypes } from './urlTypes';

export { withIosCaptchaUrlTypes };
