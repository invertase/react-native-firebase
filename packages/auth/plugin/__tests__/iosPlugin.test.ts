import path from 'path';
import { beforeEach, describe, expect, it, jest } from '@jest/globals';
import { setUrlTypesForCaptcha } from '../src/ios/urlTypes';

describe('Config Plugin iOS Tests', () => {
  beforeEach(function () {
    jest.resetAllMocks();
  });

  it('throws if path to GoogleServer-Info.plist is not provided', async () => {
    expect(() => {
      setUrlTypesForCaptcha({
        config: {
          name: 'TestName',
          slug: 'TestSlug',
          modRequest: { projectRoot: path.join(__dirname, 'fixtures') } as any,
          modResults: {},
          modRawConfig: { name: 'TestName', slug: 'TestSlug' },
          ios: {},
        },
      });
    }).toThrow(
      `[@react-native-firebase/auth] Your app.json file is missing ios.googleServicesFile. Please add this field.`,
    );
  });

  it('throws if GoogleServer-Info.plist cannot be read', async () => {
    const googleServiceFilePath = path.join(__dirname, 'fixtures', 'ThisFileDoesNotExist.plist');
    expect(() => {
      setUrlTypesForCaptcha({
        config: {
          name: 'TestName',
          slug: 'TestSlug',
          modRequest: { projectRoot: path.join(__dirname, 'fixtures') } as any,
          modResults: {},
          modRawConfig: { name: 'TestName', slug: 'TestSlug' },
          ios: { googleServicesFile: 'ThisFileDoesNotExist.plist' },
        },
      });
    }).toThrow(
      `[@react-native-firebase/auth] GoogleService-Info.plist doesn't exist in ${googleServiceFilePath}. Place it there or configure the path in app.json`,
    );
  });

  it('throws if GoogleServer-Info.plist has no reversed client id', async () => {
    expect(() => {
      setUrlTypesForCaptcha({
        config: {
          name: 'TestName',
          slug: 'TestSlug',
          modRequest: { projectRoot: path.join(__dirname, 'fixtures') } as any,
          modResults: {},
          modRawConfig: { name: 'TestName', slug: 'TestSlug' },
          ios: { googleServicesFile: 'TestGoogleService-Info.incomplete.plist' },
        },
      });
    }).toThrow(
      '[@react-native-firebase/auth] Failed to parse your GoogleService-Info.plist. Are you sure it is a valid Info.Plist file with a REVERSE_CLIENT_ID field?',
    );
  });

  it('adds url types to the Info.plist', async () => {
    const result = setUrlTypesForCaptcha({
      config: {
        name: 'TestName',
        slug: 'TestSlug',
        modRequest: { projectRoot: path.join(__dirname, 'fixtures') } as any,
        modResults: {},
        modRawConfig: { name: 'TestName', slug: 'TestSlug' },
        ios: { googleServicesFile: 'TestGoogleService-Info.plist' },
      },
    });
    expect(result.modResults).toMatchSnapshot();
  });
});
