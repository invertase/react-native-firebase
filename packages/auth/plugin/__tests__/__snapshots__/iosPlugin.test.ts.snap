// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Plugin iOS Tests adds url types to the Info.plist 1`] = `
{
  "CFBundleURLTypes": [
    {
      "CFBundleURLSchemes": [
        "com.googleusercontent.apps.SomeRandomClientIdString",
      ],
    },
  ],
}
`;
